### ISSUE TITLE
Jpa criteria support in Hibernate ORM Panache

### ISSUE THREAD

**Issue by hantsy** [quarkusio/quarkus#3965]
JPA Criteria is very important for developers. Hope add  Spring Data like `Specification` to build query Criteria and provide variants of existed methods , eg. `findAll(Specification, Sort)`.

**Comment by johnbanq** [comment]
In addition to that, can we come up with something like JHipster's entity filtering mechanism?
It allows you to write things like age.lessThan=10 in your query parameters and get it converted to criteria queries.
Not a big feature but I would say it is a nice addition

**Comment by loicmathieu** [comment]
There are ongoing discussion in this area in the following issue: https://github.com/quarkusio/quarkus/issues/2303

**Comment by emmanuelbernard** [comment]
> In addition to that, can we come up with something like JHipster's entity filtering mechanism?
> It allows you to write things like age.lessThan=10 in your query parameters and get it converted to criteria queries.
> Not a big feature but I would say it is a nice addition

If I'm not mistaken, the entity filtering is really how to expose arbitrary query restriction via your REST endpoint, correct @johnbanq? I think it's akin to the simplification of the REST en

**Comment by emmanuelbernard** [comment]
I agree with @loicmathieu I think the optional predicate will offer the same flexibility as specification and the same type safety once we plug the HQL resolver. I'd prefer we start with this approach before going on the heavyweight Specification approach Spring went to.

**Comment by hantsy** [comment]
@emmanuelbernard  yes, the flexibility is important. Another concern is the Java  **type-safe** query, which is provided by JPA Criteria API.
